db/crud.go

```diff
@@ -882,278 +1072,109 @@ func (db *Database) IsIllegalConflict(doc *Document, parentRevID string, deleted
 }
 
-// Function type for the callback passed into updateAndReturnDoc
-type updateAndReturnDocCallback func(*Document) (resultBody Body, resultAttachmentData AttachmentData, updatedExpiry *uint32, resultErr error)
+func (db *Database) documentUpdateFunc(docExists bool, doc *Document, allowImport bool, previousDocSequence uint64, unusedSequences []uint64, callback updateAndReturnDocCallback, expiry uint32) (retDoc *Document, retSyncFuncExpiry *uint32, retBody Body, retNewRevID string, retStoredBody Body, retOldBodyJSON string, retDocSequence uint64, retUnusedSequences []uint64, changedPrincipals []string, changedRoleUsers []string, err error) {
 
-// Calling updateAndReturnDoc directly allows callers to:
-//   1. Receive the updated document body in the response
-//   2. Specify the existing document body/xattr/cas, to avoid initial retrieval of the doc in cases that the current contents are already known (e.g. import).
-//      On cas failure, the document will still be reloaded from the bucket as usual.
-func (db *Database) updateAndReturnDoc(
-	docid string,
-	allowImport bool,
-	expiry uint32,
-	existingDoc *sgbucket.BucketDocument, // If existing is present, passes these to WriteUpdateWithXattr to allow bypass of initial GET
-	callback updateAndReturnDocCallback) (docOut *Document, newRevID string, err error) {
-	key := realDocID(docid)
-	if key == "" {
-		return nil, "", base.HTTPErrorf(400, "Invalid doc ID")
-	}
+	var storedBody Body
+	var syncFunctionExpiry *uint32
 
-	// Added annotation to the following variable declarations for reference during future refactoring of documentUpdateFunc into a standalone function
-	var doc *Document                                // Passed to documentUpdateFunc as pointer, may be possible to define in documentUpdateFunc
-	var body Body                                    // Could be returned by documentUpdateFunc
-	var storedBody Body                              // Persisted revision body, used to update rev cache
-	var changedPrincipals, changedRoleUsers []string // Could be returned by documentUpdateFunc
-	var docSequence uint64                           // Must be scoped outside callback, used over multiple iterations
-	var unusedSequences []uint64                     // Must be scoped outside callback, used over multiple iterations
-	var oldBodyJSON string                           // Could be returned by documentUpdateFunc.  Stores previous revision body for use by DocumentChangeEvent
-
-	// documentUpdateFunc applies the changes to the document.  Called by either WriteUpdate or WriteUpdateWithXATTR below.
-	documentUpdateFunc := func(doc *Document, docExists bool, importAllowed bool) (updatedDoc *Document, writeOpts sgbucket.WriteOptions, updatedExpiry *uint32, err error) {
+	err = db.preCallbackValidation(doc, allowImport, docExists)
+	if err != nil {
+		return doc, syncFunctionExpiry, nil, ``, storedBody, ``, previousDocSequence, unusedSequences, changedPrincipals, changedRoleUsers, err
```

Comment:
If we're returning an error, we don't care about the other return values (in all cases, as far as I can tell - the calling function is always just returning at that point).  Given the number of return values, I think it would make the code quite a bit more readable if we just did `return` here.  It's not generally good practice, but I feel like this may be an exception.  Otherwise I find myself scanning the return values to try to work out whether it's significant that we're returning `` vs retNewRevID, when really all we need to know is 'error means update is cancelled'.

Reply:
We do need to return at least docSequence, for the releaseSequence case I believe, however, may be able to work around this further down.

Reply:
Nevermind, just saw a comment further down!